system/include/emscripten/val.h

```diff
@@ -646,6 +659,104 @@ inline val::iterator val::begin() const {
   return iterator(*this);
 }
 
+#if _LIBCPP_STD_VER >= 20
+namespace internal {
+// Awaiter defines a set of well-known methods that compiler uses
+// to drive the argument of the `co_await` operator (regardless
+// of the type of the parent coroutine).
+// This one is used for Promises represented by the `val` type.
+class val_awaiter {
+  // State machine holding awaiter's current state. One of:
+  //  - initially created with promise
+  //  - waiting with a given coroutine handle
+  //  - completed with a result
+  std::variant<val, std::coroutine_handle<val::promise_type>, val> state;
+
+  constexpr static std::size_t STATE_PROMISE = 0;
+  constexpr static std::size_t STATE_CORO = 1;
+  constexpr static std::size_t STATE_RESULT = 2;
```

Comment:
AFAIK yup :(

Reply:
I mean, if you don't have duplicate types, then you can at least avoid explicit indices as it will "compute" them for you, but that wouldn't work here as 2 states both use `val`.

Reply:
Its strange they don't have version that is basically as safer/checked/tagged version of the C "union" where each member can have a name and the syntax is simple to read like the "C" union.

Reply:
Resolving this..

Reply:
> Its strange they don't have version that is basically as safer/checked/tagged version of the C "union" where each member can have a name and the syntax is simple to read like the "C" union.

Tell me about it... I guess they're hesitant to add more syntax and try to implement everything via STL, but the result is unpleasant.